### PR TITLE
NIFI-3442 - Explictly support EL on Username, password, port

### PR DESCRIPTION
**Pull request by trixpan**
Introduce basic test unit

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [X] ~~If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?~~
- [X] ~~If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?~~
- [X] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] ~~Have you ensured that format looks appropriate for the output in which it is rendered?~~

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by trixpan** [comment]
@markap14 you may want to have a look

note that in the case of transfer method I simply removed the EL evaluation. Happy to follow the opposite approach if needed

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        ftpPort = fakeFtpServer.getServerControlPort();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        fakeFtpServer.stop();
+    }
+
+    @Test
+    public void testValidators() {
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java]: This test looks... unfinished?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    }
+
+    @Test
+    public void BasicFileUpload() throws IOException {
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java]: Rename to lowercase camel case for Java consistency.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    final FakeFtpServer fakeFtpServer = new FakeFtpServer();
+    final String username = "nifi-ftp-user";
+    final String password = "Test test test chocolate";
+    int ftpPort;
+
+    @Before
+    public void setUp() throws Exception {
+        fakeFtpServer.setServerControlPort(0);
+        fakeFtpServer.addUserAccount(new UserAccount(username, password, "c:\\data"));
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java]: Is the Windows-style path cross-platform compatible?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml [hunk 4]
```diff
@@ -230,6 +230,12 @@ language governing permissions and limitations under the License. -->
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.mockftpserver</groupId>
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: I checked the [project homepage](http://mockftpserver.sourceforge.net/) to check for the license to see if `LICENSE` or `NOTICE` needed to be updated, but I don't see a license posted. Do you know which it's published under?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            attributes.put(CoreAttributes.FILENAME.key(), "randombytes-1");
+            runner.enqueue(fis, attributes);
+            runner.run();
+        }
+        runner.assertQueueEmpty();
+        runner.assertAllFlowFilesTransferred(PutFTP.REL_SUCCESS);
+        FileSystem results = fakeFtpServer.getFileSystem();
+
+        // Check file was uploaded
+        Assert.assertTrue(results.exists("c:\\data\\randombytes-1"));
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java]: Same comment as above referencing Windows-style paths.

**Comment by trixpan** [comment]
PR still not complete Andy. My bad

**Comment by alopresto** [comment]
No worries, I was just going through open PRs and trying to clean up. Ping me when you're ready.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml (hunk 4 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: ASLv2

https://sourceforge.net/p/mockftpserver/code/HEAD/tree/MockFtpServer/LICENSE.txt

Also, scope is test only. I was under the impression the NOTICE and LICENSE are only required when code is bundled as part of the NAR packaging?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java (hunk 3 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java]: should be. Work on my linux VMs. Did you have issues?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java (hunk 2 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutFTP.java]: fixed

**Comment by trixpan** [comment]
@alopresto - the missing amendments have been uploaded and hopefully also addressed the comments made earlier

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml (hunk 4 above)
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: @trixpan is correct on the LICENSE/NOTICE for binary dependencies like this.  If we're pulling in source even for test purposes though then we have to account for it.

Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: Thanks for the clarification.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFTP.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        runner.enqueue(new byte[0]);
+        pc = runner.getProcessContext();
+        if (pc instanceof MockProcessContext) {
+            results = ((MockProcessContext) pc).validate();
+        }
+        assertEquals(0, results.size());
+
+        /* Try an invalid expression */
+        results = new HashSet<>();
+        runner.setProperty(FTPTransfer.USERNAME, "$static-username}");
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFTP.java]: In the test, `MockValidationContext` is used which does not validate the actual EL expression. Therefore, I was getting a test failure here (**0** validation errors present when **1** was expected). I changed the property value to be `""` and the test succeeded (i.e. the validator threw an error).

**Comment by alopresto** [comment]
I verified contrib-check, ran the tests (as mentioned, one failed but I resolved it), and built the project. I made a sample flow which generated a flowfile, updated a `username` attribute with "anonymous", and then used a `PutFTP` processor with the **username** value having the expression `${username}`. This worked successfully. 

+1, merging. 
screen shot 2017-03-08 at 8 48 23 pm
screen shot 2017-03-08 at 8 48 08 pm

**Pull request closed without merging**